### PR TITLE
Multiproject plate: support for absquant

### PR DESCRIPTION
**Pull request by AmandaBirmingham**
Changes included here are:
* Refactor of sample_sheet to support representing absquant (and katharoseq, since I was at it) as mixins that can be added to other "base"  kinds of sample sheets.  This was necessary to limit code duplication for adding a multiproject-plate-competent absolute quant sample sheet.
        - Previously, a lot of lists with the very same contents were declared at various points in the code. For the abs-quant-related refactor, I wanted to replaced these with some constants referenced in various places, but to ensure that there was no chance of changes to a (reference to a) list in one location affecting a (reference to a) list in another place, I exchanged lists for tuples and dictionaries for MappingProxyTypes.  Thus these collection constants are shared but immutable; in the situations where a class wants to "modify" one of them, it is actually making a copy and modifying that.
* Push katharoseq support to standard_metag sheet version 102 since it isn't fully ready to use yet, make multiproject-plate-competent sample sheet into version 101.  (Also addressed a couple of little issues with katharoseq relating to col order, column checking).
* Add some more explicit checking/messaging to matrix tube notebook based on experiences (re-)running two past real-life preps through it. 
* Rename a bunch of sample sheet properties that are used only within that module with leading underscores (which is a cosmetic change but makes it a lot easier to tell which things are *supposed* to be private even though it doesn't enforce that :)
* Make a handful of tweaks to clear some FutureWarnings.

Note: This also pulls in @RodolfoSalido's changes to `master` to update the test files (NOT to include katharoseq controls and to contain values based on normalized pooling using 'Raw Reads' instead of filtered).

### CODE REVIEW

File: metapool/sample_sheet.py
Comment (antgonza) [inline metapool/sample_sheet.py:147]: I like to open all TODOs as issues so they do not get lost in the code; would that be possible? Another option could be to just make the change now. What do you think ?

File: metapool/sample_sheet.py
Comment (antgonza) [inline metapool/sample_sheet.py:190]: General question, what's the benefit of defining these variables as tuples and then transform them to lists?

File: metapool/sample_sheet.py
Comment (antgonza) [inline metapool/sample_sheet.py:643]: Excellent point and I agree on moving them to the same are of the code. Could you open an issue or address?

File: metapool/sample_sheet.py
Comment (antgonza) [inline metapool/sample_sheet.py:1676]: I don't understand this TODO. I think it's fine to fail if the Katharoseq sample-sheet does not have Katharoseq samples. If we allow this, what prevents users to simply define everything as Katharoseq sample-sheet?

File: notebooks/metagenomics_pipeline_iSeqnorm.ipynb
Comment (antgonza) [inline notebooks/metagenomics_pipeline_iSeqnorm.ipynb:1359]: This looks like an empty cell, just highlighting it in case is not actually needed.

**Review by antgonza — COMMENTED** [review]
Looks real good, thank you @AmandaBirmingham. Just some minor questions.

File: metapool/sample_sheet.py
Comment (AmandaBirmingham) [reply metapool/sample_sheet.py:147]: Sounds good, will do!  I'm deleting the TODO but leaving a brief comment in the code referencing the new issue; my reason for doing this is that, when I sit down to work on a chunk of existing code, I *don't* usually go scanning through the whole issue-base to see if any of them are suggestions relevant to that chunk that I could fix while I'm in there.  Given this, I worry about issues being where refactor suggestions go to die ;) .  Leaving a little breadcrumb to the issue will make it more likely that if I touch this code again, I remember to see if I implement its fix at that point.

File: metapool/sample_sheet.py
Comment (AmandaBirmingham) [reply metapool/sample_sheet.py:190]: An excellent question!! Made even better because I actually have an answer :D 

`CARRIED_PREP_COLUMNS` and `GENERATED_PREP_COLUMNS` are among the rare sample sheet properties that actually get accessed from outside the sample sheet object.  I have not been able to find any places where the outside users try to *set* values in these properties (which is why I felt safe turning their underlying data structures into immutable tuples) BUT some of the outside users *are* using list comprehensions on these properties.  Python doesn't have tuple comprehensions, so the outside code fails if I just have the setter return the underlying tuple.  In order to keep this refactor functionally invisible to the outside users, I am therefore making a list out of the tuple before returning it so they can keep using it the way they always have.

File: metapool/sample_sheet.py
Comment (AmandaBirmingham) [reply metapool/sample_sheet.py:1676]: Hmm, I'm confused here!  What I did in this edit was just make sure that the katharoseq functionality actually acts the way it is written to say it acts--I didn't define its intended actions myself ;)  @antgonza, looks like you approved a PR in April (https://github.com/biocore/metagenomics_pooling_notebook/pull/193) that implements *optional* katharoseq columns, so I was under the impression that maybe our end goal **was** to make every sample sheet katharoseq-capable (whether or not we actually used katharoseq for that particular experiment).  I thought this was analogous to wanting to make every sample sheet going forward multiproject-capable, whether or not we ended up putting multiple projects on the plates for in any given experiment.

Do we NOT want this?  If not, then I agree that we should make more cases fail; it will be a fairly key change to the way that katharoseq support is currently envisioned in the code, though.

File: notebooks/metagenomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [reply notebooks/metagenomics_pipeline_iSeqnorm.ipynb:1359]: Ooo, interesting!  I think this came from the stuff I back-integrated from master, since I've never actually worked on this particular notebook.  

I'd actually like to retire this notebook, because having two different metag notebooks--one of which is being maintained and one of which is *not*--available as templates for the wet lab seems like needlessly tempting fate :) Do we have any kind of procedure for retiring notebooks, or do we just delete them and let git take care of remembering what they used to be?

File: metapool/sample_sheet.py
Comment (antgonza) [reply metapool/sample_sheet.py:1676]: I see your point and maybe the way to go in the future: allow users to define whatever they need based on the samples (well, sample names) and not by the "type". If you agree with this, I'm OK with it.

File: notebooks/metagenomics_pipeline_iSeqnorm.ipynb
Comment (antgonza) [reply notebooks/metagenomics_pipeline_iSeqnorm.ipynb:1359]: Just remove, we can always bring back if necessary.

File: metapool/sample_sheet.py
Comment (AmandaBirmingham) [reply metapool/sample_sheet.py:1676]: If I understand what you are saying, I think that it doesn't require changes to the current katharoseq functioning ... do I have that right?

Comment (antgonza) [reply metapool/sample_sheet.py:1676]: I think so too.

File: metapool/sample_sheet.py
Comment (charles-cowart) [inline metapool/sample_sheet.py:1293]: Based on this definition, do we need to have MetagenomicSampleSheetv101 and MetagenomicSampleSheetv102? Because I don't believe we're run any v101 sheets through production, and the definitions below define the only difference between the two as included support for Katharoseq controls. Since katharoseq controls are optional then all v101 sheets are valid v102 sheets as well, aren't they?

File: metapool/sample_sheet.py
Comment (charles-cowart) [inline metapool/sample_sheet.py:1427]: Similar to my earlier comment regarding v101 and v102 sheets. In this case we do have some AbsQuantSampleSheetv10() sheets laying around, but since all v10 sheets would be valid v11 sheets as well, why not just add this functionality to v10?

File: metapool/sample_sheet.py
Comment (charles-cowart) [reply metapool/sample_sheet.py:1676]: Right. Since katharoseq controls were presented as an extension of BLANKs, what defines a katharoseq-enabled sample-sheet was previously the presence of sample-names beginning with 'katharo' and now their presence in the new section. It's true you still need to use one of the newer sample-sheet types in order to get the katharoseq functionality, but this is because we want to preserve the versions that we processed past sheets with.

**Review by charles-cowart — CHANGES_REQUESTED** [review]
Everything looks good! Just a few questions.

File: metapool/tests/test_amplipool.py
Comment (AmandaBirmingham) [reply metapool/tests/test_amplipool.py:167]: This is amplipool stuff, which is total greek to me, but I am pretty sure this is being created from hard-coded lists of test values during test `SetUp`:

https://github.com/biocore/metagenomics_pooling_notebook/blob/9b66a81e631c07de960678c904a7eb8830ef0b5a/metapool/tests/test_amplipool.py#L88

If you would prefer, we could drop a `dtype={'Col': 'str'}` into `pd.DataFrame`?

File: metapool/sample_sheet.py
Comment (AmandaBirmingham) [reply metapool/sample_sheet.py:1172]: Interesting ... that sounds like a fruitful refactor!

File: metapool/sample_sheet.py
Comment (AmandaBirmingham) [reply metapool/sample_sheet.py:1293]: Hahahaha, @charles-cowart, great minds think alike!  Based on the same line of thinking, I had actually killed off v102, changed v101 to derive from KLSampleSheetWithContext, and switched over all the tests and everything (which went swimmingly) ... but I then discovered that trying to run the matrix tube **notebook** to create a v101 samplesheet *with* katharoseq controls (instead of without them) caused an error due to not having all the required katharoseq columns (those `katharo_description` and `platemap_generation_date` columns that the group has been discussing on slack).  Since I couldn't get an immediate answer to what the columns should be and how they should be generated, and I needed to get this multiproject stuff into test, I backed off from the katharoseq support and put it back into a separate version.

File: metapool/sample_sheet.py
Comment (AmandaBirmingham) [reply metapool/sample_sheet.py:1427]: Well, no, v10 sample sheets are not also valid as v11 sample sheets.  You are probably remembering that there's a bunch of code in the `sample_sheet` module does not require the `SampleContext` section--but the code you might be thinking of is in `KLSampleSheet` and applies to all different versions of sample sheets.  That version-agnostic code does not require the presence of a `SampleContext` section so that it can continue to run on old sheets that don't include `SampleContext`. However, the `KLSampleSheetWithContext`-specific class definition enforces that any `KLSampleSheetWithContext`-based sheet *is* required to contain the `[SampleContext]` section header and its associated column headers line even if there are no actual records to go into that section.  

I settled on this requirement because it makes it explicit that a sheet containing a present-but-empty context section was generated by a process that knows about sample context but that no context happens to exists in this particular sheet, whereas if the context section were left out altogether in this situation, it would mean we couldn't tell by looking at a sample sheet whether it *had* no controls in it or whether we hadn't bothered to *look* for controls in it :)

File: metapool/tests/test_amplipool.py
Comment (charles-cowart) [reply metapool/tests/test_amplipool.py:167]: Ty, that sounds good.

File: notebooks/metagenomics_pipeline_iSeqnorm.ipynb
Comment (AmandaBirmingham) [reply notebooks/metagenomics_pipeline_iSeqnorm.ipynb:1359]: Great!  Currently waiting for explicit confirmation from the wet lab that they are not using the old one anymore; then I'll zap it :)

File: metapool/sample_sheet.py
Comment (charles-cowart) [reply metapool/sample_sheet.py:1285]: Just our luck! 😄 It sounds like it's doable down the road; I'll add an issue for it.

File: metapool/sample_sheet.py
Comment (charles-cowart) [reply metapool/sample_sheet.py:1419]: Very true!

File: notebooks/metagenomics_pipeline_iSeqnorm.ipynb
Comment (charles-cowart) [reply notebooks/metagenomics_pipeline_iSeqnorm.ipynb:1359]: If this is the only change I will go ahead an approve and we can merge once Antonio approves.

**Review by charles-cowart — APPROVED** [review]

**Comment by charles-cowart** [comment]
Thank you @AmandaBirmingham ! Everything looks great.

**Review by antgonza — APPROVED** [review]

**Pull request merged**
